xilem/src/view/memoize.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright 2024 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+use std::any::Any;
+
+use masonry::{widget::WidgetMut, WidgetPod};
+
+use crate::{MasonryView, MessageResult, ViewCx, ViewId};
+
+pub struct Memoize<D, F> {
+    data: D,
+    child_cb: F,
+}
+
+pub struct MemoizeState<T, A, V: MasonryView<T, A>> {
+    view: V,
+    view_state: V::ViewState,
+    dirty: bool,
+}
+
+impl<D, V, F> Memoize<D, F>
+where
+    F: Fn(&D) -> V,
+{
+    pub fn new(data: D, child_cb: F) -> Self {
+        Memoize { data, child_cb }
+    }
+}
+
+impl<State, Action, D, V, F> MasonryView<State, Action> for Memoize<D, F>
+where
+    D: PartialEq + Send + Sync + 'static,
+    V: MasonryView<State, Action>,
+    F: Fn(&D) -> V + Send + Sync + 'static,
+{
+    type ViewState = MemoizeState<State, Action, V>;
+
+    type Element = V::Element;
+
+    fn build(&self, cx: &mut ViewCx) -> (WidgetPod<Self::Element>, Self::ViewState) {
+        assert!(
+            std::mem::size_of::<F>() == 0,
+            "The callback is not allowed to be a function pointer or a closure capturing context"
+        );
```

Comment:
I think this could be turned into a compile-time check by moving it to the inherent `impl` (the one with `pub fn new`) as

```rust
const _ASSERT_CONTEXTLESS_FN: () = {
    assert!(
        std::mem::size_of::<F>() == 0,
        "The callback is not allowed to be a function pointer or a closure capturing context"
    );
};
```

Reply:
Oh pretty cool idea, I had compile-time check in mind, but thought it would be more complicated and requires unstable Rust, turns out, not the case, thanks!

I had to add `ASSERT_CONTEXTLESS_FN` as statement into the `new` function though, otherwise the compiler would just optimize it away (not sure whether that's a compiler bug, as I think it should also work without, because `const` needs to be evaluated anyways?